parallize.py

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+import argparse
+from subprocess import Popen, run
+from pathlib import Path
+
+
+def main(arguments):
+    """Takes a directory containing user programs and runs them as seperate processes
+
+    May require elevated priveges if the supplied directory is restricted (i.e. use
+    sudo when calling this)
+
+    Only runs .c, .cpp, and .py files for now
+    For .c and .cpp files:
+        Uses g++ to compile and places excecutable in program_directory/obj
+
+    Args:
+        arguments (list): list of arguments
+    """
+    parser = argparse.ArgumentParser(description=__doc__)
+
+    parser.add_argument('program_directory',
+                        help="Directory containing user programs to run",
+                        type=dir_path)
+
+    # TODO add output file for each program in a specified output directory
+
+    args = parser.parse_args(arguments)
+
+    # get directory of programs
+    program_dir = Path(args.program_directory)
+
+    # extract programs
+    py_files = list(program_dir.glob('**/*.py'))
+    c_files = list(program_dir.glob('**/*.c'))
+    cpp_files = list(program_dir.glob('**/*.cpp'))
+
+    programs = {} # program name : program id
+
+    for program in py_files:
+        pid = Popen(["python3", program]).pid
+        programs[program] = pid
+
+    if c_files or cpp_files:
+        print('Compiling source files...')
+
+    Path(program_dir, 'obj').mkdir(exist_ok=True)
+    for source_file in c_files + cpp_files:
+        program = Path(source_file.parent, 'obj', source_file.name.split('.')[0])
+        run(["g++", '-o', program, source_file])
+        pid = Popen([program]).pid
+        programs[program] = pid
+
+    if c_files or cpp_files:
+        print('Finished compilation')
+
+    print(f'Now running:')
+    for program, pid in programs.items():
+        print(f'Program: {program}, pid: {pid}')
+
+
+def dir_path(string):
+    if os.path.isdir(string):
+        return string
+    else:
+        raise NotADirectoryError(string)
+
+
```

Comment:
It seems as if this program currently kicks off all of the programs and completes (exits). This isn't exactly what I was thinking in my head but It's arguably a better solution. However, I think it would be valuable to have a way to stop all of the programs as well. I think the best solution here would be to write a separate stop programs script, but alternatively, you could add that functionality to this program. whatever you think makes the most sense I guess.

If you think this is a sufficiently large amount of work then let me know and we can just merge this and create a new issue for creating that stop script.

Reply:
The problem with that is it hangs in the shell. Since it prints all the PIDs, all you need to do is `kill <pid1> <pid2> ... <pidi>` to stop them

Reply:
So then how's the best way to handle this? I assume having two separate programs then would be easiest? We would then need a way of maintaining the running program PID's. When your fork() it should give you the PID yea? so maybe then we should write them to a running programs file for which the program "stopper" could read? unless you've got a better idea?

This extended functionality will be useful for not only stopping but pausing and further process resource allocation for future endeavors.